### ISSUE TITLE
Please don't include hidden landmarks

### ISSUE THREAD

**Issue by carmacleod** [matatk/landmarks#80]
If a landmark element is `hidden` or `display: none;` or `visibility: hidden;` please don't include it in the list of landmarks.
Maybe also `aria-hidden="true"`.

**Comment by matatk** [comment]
I think it's fair to say this is a bug; thanks for pointing it out.  Thus, I need to come up with a way to exclude elements, and have that way recognise these four different techniques for hiding content.

Not sure how I feel about `aria-hidden` at the moment: I'm struggling to think of a situation where it's justified to hide a large region of the page from users of Assistive Technologies, but not visually-browsing users.

**Comment by carmacleod** [comment]
Agreed about `aria-hidden` - I wasn't sure about it either. I'm now leaning towards keeping `aria-hidden` landmarks in the landmarks list (and consequently allowing navigation to them). You are right that page authors should be discouraged from using `aria-hidden` on landmarks.

I'm also wondering how you would detect when the `hidden`, `display: none`, or `visibility: hidden` state of a landmark changes. Do you dynamically update the list each time the user types a landmark navigation key or drops down the list?

**Comment by matatk** [comment]
ACK re `aria-hidden`.

Regarding the issue of dynamic pages, I thought about two separate situations…

* Some sites (GitHub, YouTube, …) don’t do full-page reloads, but instead change the page a lot and then use the HTML History API to make it look like the URL has changed (this allows them to produce some really nice dynamic effects and smooth UI, _and_ let users bookmark all areas).  These *are* handled by Landmarks, as per #16.

* However, pages that just make arbitrary changes to the DOM and don’t use the HTML History API, well that is not currently detected.  E.g. when a pop-up appears.  I have not looked into this too much, as it seems to get into the law of diminishing returns quite quickly.  Perhaps one way to do it would be [Mutation Observers](https://developer.mozilla.org/en/docs/Web/API/MutationObserver) but I understand this would place a significant performance burden on any site, as Landmarks would have to listen for all changes.  Further, if a pop-up appears and is coded well, then it would be both (a) simple, so not requiring landmark navigation within it and (b) constrain the focus order to its contents, so the user couldn’t accidentally fall out of it.  Both of these put tougher, plus the performance and possibly unpredictability issues have made me think this feature is not really warranted.

**Comment by carmacleod** [comment]
We have some slide-out "nav" panels - I was thinking about those.
Also, not sure if someone would want to put landmark regions in the hidden panels of an "accordion" or "tab folder" - maybe.
However, your point about performance burden and the law of diminishing returns is well made.

**Comment by matatk** [comment]
That is an interesting—and perhaps common—use case.  Also, if you have a publicly-accessible URL that you don't mind sharing, it would give me a good test case to play around with, which would be of great help in seeing if this _can_ be done in an efficient and effective way.

I sense we may be coming full-circle on this, because I just had a couple of thoughts...
* What if visually-hidden landmarks _were_ included (as they are now by virtue of what is essentially a bug) but flagged as presently hidden, so at least the user knows they're there.
* Well if we go that far, maybe we can attach a DOM mutation observer to those would-be landmark regions that can detect if they are made visible in future. So actually maybe there _is_ a way to do this efficiently in the case where the landmark already exists but is hidden.

This would leave an asymmetry whereby landmarks that are injected later will still not be picked up, but that is a bigger problem (as above), and the slide-out nav pattern is I think pretty common, and perhaps the general "have it in the DOM but hidden for now" pattern is worth supporting, as it appears there is a good way to do it.

Think I need to be looking at a couple of things:
1. Are whole-page mutation observers really slow (I gather yes, but haven't tried myself).
2. Even if whole-page observation is out: do other pages use similar "thing in the DOM that is visually-hidden for now" patterns? If so, then the above more focused technique could be a big plus, and implemented relatively tidily.

**Comment by carmacleod** [comment]
I tried both JAWS and NVDA (in Chrome). They do **not** list hidden landmarks. They don't navigate to them either. (I only tested with display: none; but I assume html hidden would be the same). So I am now leaning heavily towards our initial feeling that "If a landmark element is hidden or display: none; or visibility: hidden; please don't include it in the list of landmarks."

FYI, I also noticed that they create the list when the user requests it, because I "hid" one of the landmarks using the debugger, i.e. I added style="display: none;" to the banner, and then asked for the list of landmarks, and the banner was not in the landmarks list. Refreshing the page put the banner landmark back in the list again.

In case you would like to try this in the screen readers, here are the keyboard shortcuts:
JAWS: List Regions: INS+Ctrl+r    Next/Previous Region: r/Shift+R
NVDA: List Elements: INS+F7 (then select Landmarks radio)    Next/Previous Landmark: d/Shift+D
Note that you have to be in browse mode for these to landmark commands to work in the (Windows) screen readers. (i.e. make sure focus is not in some form element). This made me appreciate your extension even more, because it works no matter where the focus is.  :)

**Comment by matatk** [comment]
Calculating it every time would be simpler, though it would also mean that the calculation is being done a lot, and I might try to check out the alternative ideas discussed above first.

One thing I'm concerned about is, if they are re-calculated each time, how does the user keep their place in the next/previous cycle. I'm sure there are good heuristics for this, but wouldn't want to break that, so it'll need some thought.

**Comment by carmacleod** [comment]
> how does the user keep their place

Right - good point. If the current landmark disappears, then do they start over, or start at the next/previous landmark. Not sure.
I would say, though, that if the current landmark is still in the list (which would be the usual case), then they just stay there.

**Comment by matatk** [comment]
I have fixed this in the development version (the `master` branch), though am still working on the underlying issue (#84, for which I've taken on board your feedback; thanks @carmacleod).  This will be released as part of 2.0.8.

**Comment by carmacleod** [comment]
Great! Looking forward to trying it out.  :)

**Comment by matatk** [comment]
I just found a visually-apparent yet `aria-hidden` landmark in the wild, and I still agree with the above reasoning that they should be counted as a landmark by Landmarks, even though screen-readers and other assistive technologies may ignore them. In this case, it was definitely something that assistive technology users would've wanted to be able to use. (I wonder if it may've been hidden on purpose while the site developers made that part accessible, but either way, it seems helpful to surface such regions; can imagine some may contain mistakenly-included debugging information, or simply be overlooked, for example.)

In line with this stance, it seems that Landmarks should make it clear when a landmark is `aria-hidden`; it could help web authors spot potential areas of their pages that exclude assistive technology users. It feels like it could go hand-in-hand with the 'lint warnings' proposed in #182.

*Was thinking about making such indication optional, but I imagine: the indicator could be made unobtrusive for most users (e.g. it could be an icon adjacent to, or within, the landmark's button in the pop-up, with a tooltip containing more info for web authors); making it optional makes it hard to discover, and there are several options additions in the pipeline already, and I'm still aiming to keep things as simple as possible.*

Actions (to be filed in a new issue/s):
* Clarify the already-existing behaviour of showing visually-apparent but semantically-hidden landmarks in the README. Note that #85 is already referenced in the changelog section, but it's a bit confusing because it talks about "hidden" regions, as opposed to "visually-hidden" ones. It may be worth retconning this.
* Add the indication as discussed above.

**Comment by carmacleod** [comment]
> I just found a visually-apparent yet aria-hidden landmark in the wild

Wow - almost sounds like a bug.  :)

Agree that visually-apparent but aria-hidden landmarks should be in the list, and annotated. And that the feature should not be optional.  :)

What's the exact wording in the changelog? Is it "Do not show hidden landmarks"? If so, then yes, it might be less confusing if it said something like, "Do not show hidden, display:none, or visibility:hidden landmarks. Show aria-hidden landmarks with a flag."

**Comment by matatk** [comment]
I have changed my stance on the issue of [`aria-hidden`](https://www.w3.org/TR/wai-aria-1.1/#aria-hidden) (and the [proposed `inert`](https://github.com/WICG/inert)) landmark regions. I think we really have to start by going with what the content author has specified. If the author has set part of the page as semantically hidden in either of these ways, nothing inside of them should be recognised as landmark regions. This aligns with the specs. [Thanks to this comment on a follow-up issue that helped explain this to me.](https://github.com/matatk/landmarks/issues/198#issuecomment-526688662)

However, it's possible that the [proposed linting](https://github.com/matatk/landmarks/issues/182#issuecomment-418528952) could include a check for some areas being marked as semantically hidden and there _not_ being a modal dialog present. That would surface the potential author error of marking something that's actually visible and interactive as `aria-hidden="true"`.

**Comment by carmacleod** [comment]
Ok, no problem - thanks for the discussion.
People are [still arguing about whether or not aria-hidden things should be in the accessibility tree](https://github.com/w3c/aria/issues/1185). Note that that issue is closed, but people are still commenting. ;)

I think your idea to have linting is absolutely brilliant. :)

**Comment by matatk** [comment]
Thanks for the link @carmacleod; it's helpful to have context on decisions. I am following the ARIA repo now, so I can keep up with discussions.
